88 återställ lösenord reset password

joellos opened the pull request:
feat: Password Reset Implementation

Implements password reset functionality with email-based token verification.

Changes
New Endpoints:
POST /api/Auth/ForgotPassword - Request password reset email
POST /api/Auth/ResetPassword - Reset password with token

Security:
Token expiration after 24 hours (configurable)
Rate limiting: 3 requests per hour per IP
Token can only be used once
Timing attack protection prevents email enumeration
All sessions invalidated on password change

Configuration:
Added PasswordResetSettings with FrontendUrl and TokenLifetimeHours
Registered PasswordResetTokenProvider in Identity
Added PasswordResetPolicy to rate limiter

Configuration

Add to User Secrets for development:
json
{
  "PasswordResetSettings": {
    "FrontendUrl": "https://put-frontend-url"
  }
}

The FrontendUrl is where users will be directed to enter their new password.

Request Password Reset

POST /api/Auth/ForgotPassword

Request body:
json
{
  "email": "user@example.com"
}

Response: 200 OK (always returns success to prevent email enumeration)

The user receives an email with a reset link containing token and email parameters.

Reset Password

POST /api/Auth/ResetPassword

Request body:
json
{
  "email": "user@example.com",
  "token": "CfDJ8...",
  "newPassword": "NewPassword123!"
}


Success: 200 OK
Error: 400 Bad Request (invalid input) or 410 Gone (expired/invalid token)

Testing Instructions

Test Flow

1. Request password reset:
bash
POST https://localhost:7161/api/Auth/ForgotPassword
Content-Type: application/json

{
  "email": "your-test-user@example.com"
}
```

2. Check the email inbox. You will receive an email with a link like:
https://localhost:5173/reset-password?token=CfDJ8...very-long-token...&email=your-test-user@example.com

3. Extract the token from the URL. 

4. Reset the password:
bash
POST https://localhost:7161/api/Auth/ResetPassword
Content-Type: application/json

{
  "email": "your-test-user@example.com",
  "token": "full-token-here",
  "newPassword": "NewSecurePassword123!"
}

Verify the password was changed by logging in with the new password.


Testing

Test Token Reuse Prevention:
Try using the same token again. It should fail with 410 Gone.

Test Rate Limiting:
Send 4 password reset requests in quick succession. The 4th request should return 429 Too Many Requests.

Test Token Expiration:
Wait 1 hours and try to use an old token. It should fail with 410 Gone.


Password requirements:
Minimum 6 characters, at least one uppercase letter, at least one lowercase letter, at least one number and at least one special character


Security Notes:

Tokens are single-use and expire after 1 hour
All user sessions are terminated when password is changed
Rate limiting prevents brute force attacks
Timing attack protection prevents user enumeration
Always returns 200 OK for forgot password regardless of user existence

Session Invalidation Behavior:

When a user resets their password:
UpdateSecurityStamp invalidates all existing JWT tokens
InvalidateRefreshTokenAsync clears all refresh tokens
User is logged out from all devices
User must log in again with new password

This is a security feature to ensure compromised sessions cannot continue after password change.

joellos committed:
Added request password reset and confirm password reset. Will fix URL decode

joellos committed:
updated service to accept decoded token as well

joellos committed:
updated validators with proper comments

joellos committed:
added proper attributes

joellos committed:
added helper file for the tokens for reset password since the identity tokens are hardcoded and I don't want to touch the existing tokens lifetimes. And made the passwordResetSettings configurable from appsettings from appsettings to follow the existing pattern and code structure and to make it easy to change.

joellos committed:
Merged developer into branch

joellos committed:
fixed ServiceExtension

joellos committed:
feat: implement password reset functionality

Add complete password reset flow with email verification:
- Generate password reset tokens with 24h expiration
- Send reset links via Mailjet
- Validate tokens and update passwords
- Rate limit reset requests (3/hour)
- Add startup validation for required configuration

Technical changes:
- Register PasswordResetTokenProvider in Identity
- Add PasswordResetSettings configuration
- Implement ForgotPassword and ResetPassword endpoints
- Add dedicated rate limiting policy
- Add token decoding validation
- Add timing attack protection

Breaking changes:
- Requires PasswordResetSettings:FrontendUrl in configuration

joellos committed:
Added token validation check before password reset
  - Tokens are now properly verified before allowing password change
  - Expired or invalid tokens return 410 Gone
  - Prevents token reuse attacks
- Security stamp updated BEFORE success response to ensure token invalidation

joellos committed:
added logger and fixed return responses

joellos committed:
merged autocomplete into branch

joellos committed:
moved code

OssyCA commented on DTOs/AuthDTOs/ConfirmPasswordResetDto.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace PegasusBackend.DTOs.AuthDTOs
+{
+    public class ConfirmPasswordResetDto
```
If we are using Fluent validator, we dont need DataAnnotations.

joellos replied:
ah shit, thats my bad, i have validator but i followed a youtube guide when implementing in the beginning, i will fix that asap

joellos committed:
removed data annotations

joellos committed:
removed redundancy and changed to bool instead of string for better consistency with resetpassword

joellos committed:
updated to correct rate limit policy

joellos committed:
opps

joellos commented on Services/PasswordResetService.cs:
```diff
@@ -0,0 +1,143 @@
+            _logger = logger;
+        }
+
+        public async Task<ServiceResponse<string>> ForgotPasswordAsync(RequestPasswordResetDto request)
```
okay, thank you for pointing that out. I will fix this

joellos committed:
fix patch timing attack and race condition in password reset

make response time constant (800 ms) to prevent email enumeration

replace remove+Add with resetPasswordAsync

added rate limiting to ResetPassword endpoint

sanitize error messages to avoid leaking technical details

joellos committed:
inconsistency fixes, url fix, made sure only the latest reset token works

joellos committed:
moved passwordsettings to only appsettings

joellos committed:
Merge branch 'Developer' into 88-återställ-lösenord-reset-password

Marremelad reviewed (approved):
LGTM!
